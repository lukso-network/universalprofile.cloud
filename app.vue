<script setup lang="ts">
import { fetchProfile } from '@/utils/fetchProfile'
import { PROVIDERS } from '@/types/enums'

if (typeof window !== 'undefined') {
  import('@lukso/web-components')
}

// setup translations
useIntl().setupIntl(defaultConfig)

// setup web3 instances
const web3Store = useWeb3Store()
const appStore = useAppStore()
const provider = window.ethereum
if (provider) {
  web3Store.addWeb3(PROVIDERS.INJECTED, provider) // for chain interactions through wallet
}
web3Store.addWeb3(
  PROVIDERS.RPC,
  appStore.getNetwork(appStore.selectedNetwork).rpcHttp
) // for chain interactions through RPC endpoint

onMounted(async () => {
  const { setLoading } = useProfileStore()
  try {
    setLoading(true)
    // fetching profile in root as it's used in every page
    await fetchProfile()
  } catch (error) {
    console.error(error)
    // TODO redirect to 404 page once it's added
  } finally {
    setLoading(false)
  }
})
</script>

<template>
  <div>
    <NuxtLayout>
      <NuxtPage />
    </NuxtLayout>
  </div>
</template>
