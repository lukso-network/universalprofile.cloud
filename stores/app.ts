import type { Modal } from '@/types/modal'
import type { NetworkInfo, NetworkId } from '@/types/network'

/**
 * App store
 * Keeps the information about app
 *
 */
export const useAppStore = defineStore(
  'app',
  () => {
    const networkConfig = useNetworkConfig()
    const networks = ref<NetworkInfo[]>(networkConfig)
    const selectedChainId = ref<string>(DEFAULT_NETWORK_CHAIN_ID)
    const modal = ref<Modal>()
    const connectedProfileAddress = ref<Address>()
    const assetFilter = ref(AssetFilter.owned)
    const hasSimpleNavbar = ref<boolean>(false)

    // statuses
    const isConnecting = ref(false)
    const isLoadingProfile = ref(false)
    const isLoadingAssets = ref(false)
    const isLoadedApp = ref(false)
    const isSearchOpen = ref(false)

    // --- getters

    const getNetworkByChainId = (chainId: string): NetworkInfo => {
      const network = networks.value.find(
        network => network.chainId === chainId
      )

      // fallback to default network
      if (!network) {
        return networks.value.find(
          network => network.chainId === DEFAULT_NETWORK_CHAIN_ID
        ) as NetworkInfo
      }

      return network
    }

    const getNetworkById = (id: NetworkId): NetworkInfo => {
      const network = networks.value.find(network => network.id === id)

      if (!network) {
        return networks.value.find(
          network => network.chainId === DEFAULT_NETWORK_CHAIN_ID
        ) as NetworkInfo
      }

      return network
    }

    const currentNetwork = computed(() =>
      getNetworkByChainId(selectedChainId.value)
    )

    const isConnected = computed(() => !!connectedProfileAddress.value)

    const isTestnet = computed(() => selectedChainId.value === TESTNET_CHAIN_ID)

    // --- actions

    const setModal = (newModal: Modal) => {
      modal.value = newModal
    }

    return {
      networks,
      selectedChainId,
      getNetworkByChainId,
      getNetworkById,
      modal,
      setModal,
      currentNetwork,
      connectedProfileAddress,
      isConnected,
      isConnecting,
      isLoadingProfile,
      isLoadingAssets,
      isLoadedApp,
      assetFilter,
      isTestnet,
      hasSimpleNavbar,
      isSearchOpen,
    }
  },
  {
    persist: {
      paths: ['connectedProfileAddress', 'assetFilter', 'selectedChainId'],
      key: STORAGE_KEY.APP_STORE,
    },
  }
)
