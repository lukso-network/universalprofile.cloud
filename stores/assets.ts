export const useAssetsStore = defineStore('assets', () => {
  const lsp7Assets = ref<Lsp7AssetType[]>()
  const lsp8Assets = ref<Lsp8AssetType[]>()

  const setLsp7Assets = (newAssets: Lsp7AssetType[]) => {
    lsp7Assets.value = newAssets
  }

  const setLsp8Assets = (newAssets: Lsp8AssetType[]) => {
    lsp8Assets.value = newAssets
  }

  return {
    lsp7Assets,
    lsp8Assets,
    setLsp7Assets,
    setLsp8Assets,
  }
})
