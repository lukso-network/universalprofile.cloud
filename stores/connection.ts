/**
 * Connection store
 * Keeps the information about connected profile
 *
 */
export const useConnectionStore = defineStore('connection', () => {
  const profile = reactive<Profile>({} as Profile)
  const status = reactive({ isConnected: false, isProfileLoading: false })

  // --- actions

  const setConnectedAddress = (newAddress: Address) => {
    profile.address = newAddress
  }

  const setConnectedProfile = (newProfile: Profile) => {
    Object.assign(profile, newProfile)
  }

  const clearConnectedProfile = () => {
    Object.assign(profile, {})
  }

  const setIsConnected = (value: boolean) => {
    setStatus('isConnected', value)
  }

  const setStatus = (statusName: keyof typeof status, newStatus: boolean) => {
    status[statusName] = newStatus
  }

  const reloadConnectedProfile = (address: Address, profile: Profile) => {
    clearConnectedProfile()
    setConnectedAddress(address)
    setConnectedProfile(profile)
  }

  return {
    setConnectedAddress,
    profile,
    setConnectedProfile,
    status,
    clearConnectedProfile,
    setIsConnected,
    reloadConnectedProfile,
    setStatus,
  }
})
