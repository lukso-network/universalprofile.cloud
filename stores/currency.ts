import { CurrencyList, CurrencySymbol } from '@/types/currency'

/**
 * Currency store
 * Keeps the information about currency settings
 * Data is stored in local storage
 *
 */
export const useCurrencyStore = defineStore(
  'currency',
  () => {
    const currencyList = ref<CurrencyList>()
    const currentCurrencySymbol = ref<CurrencySymbol>(DEFAULT_CURRENCY_SYMBOL)

    const getCurrencyMultiplier = () => {
      return (symbol: string) =>
        currencyList.value &&
        currencyList.value[symbol] &&
        currencyList.value[symbol][currentCurrencySymbol.value]
    }

    return {
      currentCurrencySymbol,
      currencyList,
      getCurrencyMultiplier,
    }
  },
  {
    persist: {
      key: STORAGE_KEY.CURRENCY_STORE,
    },
  }
)
