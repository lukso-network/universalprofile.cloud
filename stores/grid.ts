export const useGridStore = defineStore(
  'grid',
  () => {
    const { connectedProfileAddress } = storeToRefs(useAppStore())

    const isEditingGrid = ref(false)
    const hasUnsavedGrid = ref(false)
    const viewedGrid = ref<Grid[]>([])
    const isSavingGrid = ref(false)
    const selectedGridId = ref<string | undefined>()
    const gridRowHeightRatio = ref(DEFAULT_GRID_ROW_HEIGHT_RATIO)
    const gridChainId = ref<string>(DEFAULT_NETWORK_CHAIN_ID)
    const tempGrids = ref<Record<string, Grid[]>>({})
    const mobileLimitationsDisplayed = ref(false)

    // We use tempGrid as a proxy to the actual grid data stored in tempGrids
    const tempGrid = computed<Grid[]>({
      get: () => {
        if (!connectedProfileAddress.value) {
          return []
        }

        return (
          tempGrids.value[connectedProfileAddress.value.toLowerCase()] || []
        )
      },
      set: value => {
        if (!connectedProfileAddress.value) {
          return
        }

        tempGrids.value = {
          ...tempGrids.value,
          [connectedProfileAddress.value.toLowerCase()]: value,
        }
      },
    })

    return {
      isEditingGrid,
      hasUnsavedGrid,
      viewedGrid,
      tempGrid,
      tempGrids,
      isSavingGrid,
      selectedGridId,
      gridRowHeightRatio,
      gridChainId,
      mobileLimitationsDisplayed,
    }
  },
  {
    persist: {
      paths: [
        'isEditingGrid',
        'hasUnsavedGrid',
        'selectedGridId',
        'gridRowHeightRatio',
        'gridChainId',
        'mobileLimitationsDisplayed',
        'tempGrids',
      ],
      key: STORAGE_KEY.GRID_STORE,
    },
  }
)
