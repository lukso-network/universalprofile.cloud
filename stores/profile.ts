export const useProfileStore = defineStore('profile', () => {
  const address = ref<Address>()
  const name = ref<string>()
  const backgroundUrl = ref<string>()
  const profileUrl = ref<string>()

  const setAddress = (newAddress: Address) => {
    address.value = newAddress
  }

  return {
    address,
    setAddress,
    name,
    backgroundUrl,
    profileUrl,
  }
})
