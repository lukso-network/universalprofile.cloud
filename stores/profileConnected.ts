import { Profile } from '@/types/profile'

/**
 * Connected profile store
 *
 * Keeps the information about connected profile
 *
 */
export const useConnectedProfileStore = defineStore('profileConnected', () => {
  const profile = reactive<Profile>({} as Profile)
  const status = reactive({
    isConnected: false,
    isProfileLoading: false,
    isProfileLoaded: false,
    isConnecting: false,
  })

  // --- actions

  const setStatus = (statusName: keyof typeof status, newStatus: boolean) => {
    status[statusName] = newStatus
  }

  return {
    ...useProfileBase(profile),
    profile,
    status,
    setStatus,
  }
})
