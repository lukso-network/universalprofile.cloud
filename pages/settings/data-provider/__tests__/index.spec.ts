import { renderSuspended } from '@nuxt/test-utils/runtime'
import { createTestingPinia } from '@pinia/testing'
import { beforeEach, describe, expect, it, vi } from 'vitest'

import { useAppStore } from '@/stores/app'
import Index from '../index.vue'

describe('Settings data-provider page', () => {
  beforeEach(() => {
    useIntl().setupIntl(defaultConfig)
  })

  it('should render with centralized setting', async () => {
    const component = await renderSuspended(Index, {
      global: {
        plugins: [createTestingPinia()],
      },
    })

    expect(component.html()).toMatchSnapshot()
  })

  it('should render with decentralized setting', async () => {
    const component = await renderSuspended(Index, {
      global: {
        plugins: [
          createTestingPinia({
            initialState: {
              app: { fetchDataProvider: 'rpc' },
            },
          }),
        ],
      },
    })

    expect(component.html()).toMatchSnapshot()
  })

  it('can change setting', async () => {
    const component = await renderSuspended(Index, {
      global: {
        plugins: [createTestingPinia()],
      },
    })

    expect(component.html()).toMatchSnapshot()
    await component.getByText('Read from smart contracts').click()
    expect(component.html()).toMatchSnapshot()
  })
})
