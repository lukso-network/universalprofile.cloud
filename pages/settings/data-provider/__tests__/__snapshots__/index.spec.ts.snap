// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Settings data-provider page > can change setting 1`] = `
"<div id="test-wrapper">
  <div class="relative">
    <div class="relative px-4 py-6 sm:py-20 animate-fade-in opacity-100 transition-opacity delay-200  duration-300">
      <div class="mx-auto w-full sm:w-[440px]">
        <h1 class="heading-inter-17-semi-bold group mb-8 flex cursor-pointer items-center gap-1 border-b border-b-neutral-90 pb-6">
          <lukso-icon name="arrow-left-lg" class="transition group-hover:-translate-x-1"></lukso-icon> Application endpoint
        </h1><!-- Enable GraphQL data provider -->
        <section class="group w-full select-none border border-b-0 border-neutral-90 bg-neutral-100 p-4 transition first-of-type:rounded-t-12 last-of-type:rounded-b-12 last-of-type:border-b cursor-pointer">
          <div class="grid grid-cols-[max-content,auto] gap-2">
            <div>
              <div>
                <lukso-icon name="empty-outline" color="neutral-90" class="inline-flex group-hover:hidden"></lukso-icon>
                <lukso-icon name="empty-outline" color="neutral-80" class="hidden group-hover:inline-flex"></lukso-icon>
              </div>
            </div>
            <div class="paragraph-inter-16-semi-bold">
              <h3 class="pb-1">Centralised</h3>
              <p class="paragraph-inter-14-regular text-neutral-40">Load data via the LUKSO indexing service. This results in faster loading times and smoother UX, however offensive or inappropriate content can be removed.</p>
            </div>
          </div>
        </section><!-- Enable RPC data provider -->
        <section class="group w-full select-none border border-b-0 border-neutral-90 bg-neutral-100 p-4 transition first-of-type:rounded-t-12 last-of-type:rounded-b-12 last-of-type:border-b">
          <div class="grid grid-cols-[max-content,auto] gap-2">
            <div>
              <lukso-icon name="complete-filled" color="green-54" secondary-color="neutral-100"></lukso-icon>
            </div>
            <div class="paragraph-inter-16-semi-bold">
              <h3 class="pb-1">Decentralised</h3>
              <p class="paragraph-inter-14-regular text-neutral-40">Load data directly from the blockchain. This can result in delays during the loading of assets, and is censorship resistant. Some app features might be limited.</p>
            </div>
          </div>
        </section>
      </div>
    </div>
    <!--v-if-->
  </div>
</div>"
`;

exports[`Settings data-provider page > can change setting 2`] = `
"<div id="test-wrapper">
  <div class="relative">
    <div class="relative px-4 py-6 sm:py-20 animate-fade-in opacity-100 transition-opacity delay-200  duration-300">
      <div class="mx-auto w-full sm:w-[440px]">
        <h1 class="heading-inter-17-semi-bold group mb-8 flex cursor-pointer items-center gap-1 border-b border-b-neutral-90 pb-6">
          <lukso-icon name="arrow-left-lg" class="transition group-hover:-translate-x-1"></lukso-icon> Application endpoint
        </h1><!-- Enable GraphQL data provider -->
        <section class="group w-full select-none border border-b-0 border-neutral-90 bg-neutral-100 p-4 transition first-of-type:rounded-t-12 last-of-type:rounded-b-12 last-of-type:border-b cursor-pointer">
          <div class="grid grid-cols-[max-content,auto] gap-2">
            <div>
              <div>
                <lukso-icon name="empty-outline" color="neutral-90" class="inline-flex group-hover:hidden"></lukso-icon>
                <lukso-icon name="empty-outline" color="neutral-80" class="hidden group-hover:inline-flex"></lukso-icon>
              </div>
            </div>
            <div class="paragraph-inter-16-semi-bold">
              <h3 class="pb-1">Centralised</h3>
              <p class="paragraph-inter-14-regular text-neutral-40">Load data via the LUKSO indexing service. This results in faster loading times and smoother UX, however offensive or inappropriate content can be removed.</p>
            </div>
          </div>
        </section><!-- Enable RPC data provider -->
        <section class="group w-full select-none border border-b-0 border-neutral-90 bg-neutral-100 p-4 transition first-of-type:rounded-t-12 last-of-type:rounded-b-12 last-of-type:border-b">
          <div class="grid grid-cols-[max-content,auto] gap-2">
            <div>
              <lukso-icon name="complete-filled" color="green-54" secondary-color="neutral-100"></lukso-icon>
            </div>
            <div class="paragraph-inter-16-semi-bold">
              <h3 class="pb-1">Decentralised</h3>
              <p class="paragraph-inter-14-regular text-neutral-40">Load data directly from the blockchain. This can result in delays during the loading of assets, and is censorship resistant. Some app features might be limited.</p>
            </div>
          </div>
        </section>
      </div>
    </div>
    <!--v-if-->
  </div>
</div>"
`;

exports[`Settings data-provider page > should render with centralized setting 1`] = `
"<div id="test-wrapper">
  <div class="relative">
    <div class="relative px-4 py-6 sm:py-20 animate-fade-in opacity-100 transition-opacity delay-200  duration-300">
      <div class="mx-auto w-full sm:w-[440px]">
        <h1 class="heading-inter-17-semi-bold group mb-8 flex cursor-pointer items-center gap-1 border-b border-b-neutral-90 pb-6">
          <lukso-icon name="arrow-left-lg" class="transition group-hover:-translate-x-1"></lukso-icon> Application endpoint
        </h1><!-- Enable GraphQL data provider -->
        <section class="group w-full select-none border border-b-0 border-neutral-90 bg-neutral-100 p-4 transition first-of-type:rounded-t-12 last-of-type:rounded-b-12 last-of-type:border-b cursor-pointer">
          <div class="grid grid-cols-[max-content,auto] gap-2">
            <div>
              <div>
                <lukso-icon name="empty-outline" color="neutral-90" class="inline-flex group-hover:hidden"></lukso-icon>
                <lukso-icon name="empty-outline" color="neutral-80" class="hidden group-hover:inline-flex"></lukso-icon>
              </div>
            </div>
            <div class="paragraph-inter-16-semi-bold">
              <h3 class="pb-1">Centralised</h3>
              <p class="paragraph-inter-14-regular text-neutral-40">Load data via the LUKSO indexing service. This results in faster loading times and smoother UX, however offensive or inappropriate content can be removed.</p>
            </div>
          </div>
        </section><!-- Enable RPC data provider -->
        <section class="group w-full select-none border border-b-0 border-neutral-90 bg-neutral-100 p-4 transition first-of-type:rounded-t-12 last-of-type:rounded-b-12 last-of-type:border-b">
          <div class="grid grid-cols-[max-content,auto] gap-2">
            <div>
              <lukso-icon name="complete-filled" color="green-54" secondary-color="neutral-100"></lukso-icon>
            </div>
            <div class="paragraph-inter-16-semi-bold">
              <h3 class="pb-1">Decentralised</h3>
              <p class="paragraph-inter-14-regular text-neutral-40">Load data directly from the blockchain. This can result in delays during the loading of assets, and is censorship resistant. Some app features might be limited.</p>
            </div>
          </div>
        </section>
      </div>
    </div>
    <!--v-if-->
  </div>
</div>"
`;

exports[`Settings data-provider page > should render with decentralized setting 1`] = `
"<div id="test-wrapper">
  <div class="relative">
    <div class="relative px-4 py-6 sm:py-20 animate-fade-in opacity-100 transition-opacity delay-200  duration-300">
      <div class="mx-auto w-full sm:w-[440px]">
        <h1 class="heading-inter-17-semi-bold group mb-8 flex cursor-pointer items-center gap-1 border-b border-b-neutral-90 pb-6">
          <lukso-icon name="arrow-left-lg" class="transition group-hover:-translate-x-1"></lukso-icon> Application endpoint
        </h1><!-- Enable GraphQL data provider -->
        <section class="group w-full select-none border border-b-0 border-neutral-90 bg-neutral-100 p-4 transition first-of-type:rounded-t-12 last-of-type:rounded-b-12 last-of-type:border-b cursor-pointer">
          <div class="grid grid-cols-[max-content,auto] gap-2">
            <div>
              <div>
                <lukso-icon name="empty-outline" color="neutral-90" class="inline-flex group-hover:hidden"></lukso-icon>
                <lukso-icon name="empty-outline" color="neutral-80" class="hidden group-hover:inline-flex"></lukso-icon>
              </div>
            </div>
            <div class="paragraph-inter-16-semi-bold">
              <h3 class="pb-1">Centralised</h3>
              <p class="paragraph-inter-14-regular text-neutral-40">Load data via the LUKSO indexing service. This results in faster loading times and smoother UX, however offensive or inappropriate content can be removed.</p>
            </div>
          </div>
        </section><!-- Enable RPC data provider -->
        <section class="group w-full select-none border border-b-0 border-neutral-90 bg-neutral-100 p-4 transition first-of-type:rounded-t-12 last-of-type:rounded-b-12 last-of-type:border-b">
          <div class="grid grid-cols-[max-content,auto] gap-2">
            <div>
              <lukso-icon name="complete-filled" color="green-54" secondary-color="neutral-100"></lukso-icon>
            </div>
            <div class="paragraph-inter-16-semi-bold">
              <h3 class="pb-1">Decentralised</h3>
              <p class="paragraph-inter-14-regular text-neutral-40">Load data directly from the blockchain. This can result in delays during the loading of assets, and is censorship resistant. Some app features might be limited.</p>
            </div>
          </div>
        </section>
      </div>
    </div>
    <!--v-if-->
  </div>
</div>"
`;
